Feat/strict null checks

guor8lei opened the pull request:
Fixes #10 and #6 

Takes care of most cases for strictNullChecks. 

Basic pseudocode for fix: 
- For objects that are possibly null/undefined, adds the definite assignment assertion (e.g. `v.toString();` -> `v!.toString();` if `v` is possibly null/undefined)
- For variables, parameters, or properties that are assigned to null or undefined, expands their declaration to include those assigned types (e.g. `function foo(a: number)` -> `function foo(a: number|null)` if `a` is assigned to null in `foo`)
- For arguments that don't match the parameter types of functions that they are passed into, adds the definite assignment assertion (e.g. `Math.abs(n)` -> `Math.abs(n!)` if `n` is possibly null/undefined).

Cases not taken care of:
- Pushing values into an empty array, couldn't fix because diagnostic isn't emitted by the compiler.
- For inline functions, as of now it just automatically adds both `null` and `undefined` types to the parameter, I couldn't think of a smart way to infer what exact types it should expand to.

guor8lei commented on src/error_detectors/prod_error_detector.ts:
```diff
@@ -88,23 +88,21 @@ export class ProdErrorDetector extends ErrorDetector {
         });
       });
     }
```
I moved the reverse functionality to the other function because I noticed that in some particular cases strictNullChecks, sometimes the error can appear on too many possible node types for me to take account of, so I had to filter out nodes by length instead of kind in those particular cases. However, the reverse functionality was needed whether or not I filtered out by kind or not.

jiady commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+            if (
+              errorNode.getChildCount() === 1 ||
+              (errorNode.getChildCount() === 2 &&
+                errorNode.getLastChildOrThrow().getKind() ===
```
shall we just skip this case if lastChild throw error?

guor8lei replied:
Inserted guards to skip.

jiady commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+
+    // Expand all declarations to include the calculated set of types
+    modifiedDeclarationTypes.forEach((types, declaration) => {
+      const newDeclaration = declaration.setType(
```
shall we sort modifiedDeclarationTypes's value?

jiady commented on src/types.ts:
```diff
@@ -40,4 +40,9 @@ export type NodeDiagnostic = {
 
 export enum ErrorCodes {
```
let's add some documents to point out where these numbers come from.

michaelyara commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+        // When two types are not assignable to each other, add both types to the declaration (variable,
+        // parameter, or property declaration)
+        case ErrorCodes.TypeANotAssignableToTypeB: {
+          if (!this.nodeKinds.has(errorNode.getKind())) {
```
+1 for Guard Clauses!

michaelyara commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+              errorNode = errorNode.replaceWithText('return undefined!;');
+            } else {
+              errorNode = errorNode.replaceWithText(
+                'return (' + errorNode.getChildAtIndex(1).getText() + ')!;'
```
You may want o use string formatting here instead of the concatenation.

guor8lei replied:
Good point! Fixed

michaelyara commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+            }
+
+            modifiedStatementedNodes.add([
+              (errorNode.getParentOrThrow() as unknown) as StatementedNode,
```
Why do you need the double cast?

guor8lei replied:
I ran into the following error when I tried doing a direct cast:

`Conversion of type 'Node<Node>' to type 'StatementedNode' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.`

Seems like TypeScript has some extra guards for protecting against potentially dangerous casts. One way to get around this is assigning the parent node to a variable and doing a type guard, so I'll try doing that instead.

michaelyara commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+          .map(type => {
+            return type instanceof Type ? type.getText() : type;
+          })
+          .join(' | ')
```
Joining a Join. I love it!

guor8lei commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+            const modifiedStatement = this.getModifiedStatement(newNode);
+
+            modifiedStatementedNodes.add([
+              (modifiedStatement.getParentOrThrow() as unknown) as StatementedNode,
```
Inserted type guards so that it'll skip the fix if there is no parent.

guor8lei commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +37,270 @@ export class StrictNullChecksManipulator extends Manipulator {
+    errorNodes.forEach(({node: errorNode, diagnostic: diagnostic}) => {
+      const parent = errorNode.getParentOrThrow();
+
+      switch (diagnostic.getCode()) {
```
Thanks for the rec!

guor8lei commented:
Fixed for all the review comments! I also added additional fixes that takes care of certain edge cases, especially when instantiating an empty array (it's default type is `never[]` which makes it impossible to perform any operations on the array with the `strictNullChecks` flag on. Right now, I explicitly change the type to `any[]` and will try to further address this to avoid `any` when addressing `noImplicitAny`.

guor8lei commented:
I plan on including the golden files soon!

jiady commented on src/manipulators/strict_null_checks_manipulator.ts:
```diff
@@ -26,9 +41,488 @@ export class StrictNullChecksManipulator extends Manipulator {
+   * @param {DeclarationType} modifiedDeclarationTypes - Map of expanded declaration types.
+   * @param {Set<[StatementedNode, number]>} modifiedStatementedNodes - Set of modified statements.
+   */
+  private handleNonAssignableArgumentTypes(
```
let's inline code examples to fix into the code block or if branches.

guor8lei commented on tsconfig.json:
```diff
@@ -11,7 +11,8 @@
     "strictPropertyInitialization": true,
-    "noImplicitReturns": true
+    "noImplicitReturns": true,
+    "sourceMap": true
```
Fixed

guor8lei commented on test/test_files/strict_null_checks_sample.ts:
```diff
@@ -0,0 +1,132 @@
+  const n: number | null = null;
+  // Fix: return n!;
+  return n;
+}
```
Added test case

guor8lei commented on test/test_files/strict_null_checks_sample.ts:
```diff
@@ -0,0 +1,132 @@
+interface BasicInterface {}
```
Added golden file and Jasmine test -> will split the file into smaller, separate files soon!

jiady commented on test/strict_null_checks_test.ts:
```diff
@@ -0,0 +1,46 @@
+    project.resolveSourceFileDependencies();
+
+    new Runner(
+      undefined,
```
let's add comments:
/*varName*/undefined. 

if it is not readable

jiady commented on test/test_files/expected_output/strict_null_checks_sample.ts:
```diff
@@ -0,0 +1,165 @@
+interface BasicInterface {}
```
let's put folder name as test_files/golden/strict_null_checks_sample.ts

jiady commented on test/source_file_matcher.ts:
```diff
@@ -0,0 +1,74 @@
+import MatchersUtil = jasmine.MatchersUtil;
```
let's rename as testing/source_file_matcher.ts 

usually testing/xxx means helper or util for tests.
test/xxx means test cases.
